### PR TITLE
Feature/rviz plugin

### PR DESCRIPTION
**Pull request by ZacharyTaylor**
The current marker array we use for meshing needs to send the entire mesh every update. It also sends the color as 4 floats. This combines to mean meshing of even moderately large areas is impossible to do in real time. This plugin makes use of the incremeantal updates to only change the voxels that need it and also transfers only what is needed to do the rendering. This allows much higher render rates and larger renders.

Surfaces seen from the back are also invisible so you can see through ceilings without having to cut them out (originally a bug, but I think it actually improves things).

[example](http://i.imgur.com/aOlBdWw.gifv)

### CODE REVIEW

**Comment by ZacharyTaylor** [comment]
Anyone know how to make jenkins rebuild its messages?

File: voxblox_ros/src/tsdf_server.cc [hunk 1]
```diff
@@ -313,11 +310,10 @@ bool TsdfServer::generateMesh() {
   generate_mesh_timer.Stop();
 
   timing::Timer publish_mesh_timer("mesh/publish");
-  visualization_msgs::MarkerArray marker_array;
-  marker_array.markers.resize(1);
-  fillMarkerWithMesh(mesh_layer_, color_mode_, &marker_array.markers[0]);
-  marker_array.markers[0].header.frame_id = world_frame_;
-  mesh_pub_.publish(marker_array);
+  voxblox_msgs::Mesh mesh_msg;
+  fillVoxbloxMeshWithMesh(mesh_layer_, color_mode_, &mesh_msg);
```
Comment (mfehr) [inline voxblox_ros/src/tsdf_server.cc]: nit: name is a bit strange, maybe convert/generateToVoxbloxMeshMsg?

File: voxblox_rviz_plugin/src/voxblox_mesh_visual.cc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    } else {
+      std::string object_name = std::to_string(index.x()) + std::string(" ") +
+                                std::to_string(index.y()) + std::string(" ") +
+                                std::to_string(index.z());
+      ogre_object = scene_manager_->createManualObject(object_name);
+      object_map_.insert(std::make_pair(index, ogre_object));
+
+      frame_node_->attachObject(ogre_object);
+    }
+
```
Comment (mfehr) [inline voxblox_rviz_plugin/src/voxblox_mesh_visual.cc]: check if ogre_object not null?

File: voxblox_ros/include/voxblox_ros/mesh_vis.h [hunk 3]
```diff
@@ -98,6 +99,79 @@ inline void heightColorFromVertex(const Point& vertex,
...
+            break;
+          case kNormals:
+            normalColorFromNormal(mesh->normals[i + j], &color_msg);
+            break;
+          case kLambert:
+            lambertColorFromNormal(mesh->normals[i + j], &color_msg);
+            break;
+          case kLambertColor:
+            lambertColorFromColorAndNormal(mesh->colors[i + j],
+                                           mesh->normals[i], &color_msg);
```
[the remark this answers is not in the document]
Comment (ZacharyTaylor) [reply voxblox_ros/include/voxblox_ros/mesh_vis.h]: nope, good catch

File: voxblox_ros/src/tsdf_server.cc (hunk 1 above)
Comment (ZacharyTaylor) [reply voxblox_ros/src/tsdf_server.cc]: The other functions that do similar things are all called fill____WithMesh, though I agree it sounds weird. Will rename

**Pull request merged**